File: sdks/arc-go-sdk/arc.go
// Package arc is the official Go client for arc's Engine-A (infrastructure secrets) HTTP
// API. It targets the same surface the operator + agent use: log in via an auth method,
// read KV v2 secrets, issue dynamic credentials, and manage leases.
//
// It is deliberately dependency-free (standard library only) so it drops into any Go
// service without dragging in a transitive graph. Engine-B (the end-to-end vault) is NOT
// exposed here — that's a client-side zero-knowledge cryptosystem; use the TypeScript SDK
// for vault items.
//
// Trust model: this client never decides authorization. arc-server's JwtAuthGuard +
// CapabilityGuard (@arc/grants) gate every call; a 403 is returned as an *APIError without
// retry. On a 401 the client drops its cached token and retries the request once.
package arc

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// Client is a handle to an arc-server instance. Construct it with New and (optionally)
// authenticate via LoginKubernetes or SetToken. Safe for concurrent use.
type Client struct {
	baseURL    string
	httpClient *http.Client

	mu    sync.Mutex
	token string
}

// Option configures a Client.
type Option func(*Client)

// WithHTTPClient overrides the default *http.Client (useful for custom transports / tests).
func WithHTTPClient(h *http.Client) Option {
	return func(c *Client) { c.httpClient = h }
}

// WithToken seeds the client with an existing arc JWT (e.g. obtained out of band).
func WithToken(token string) Option {
	return func(c *Client) { c.token = token }
}

// New returns a Client for the arc-server at baseURL (e.g. "https://arc.svc:3001").
func New(baseURL string, opts ...Option) *Client {
	c := &Client{
		baseURL:    strings.TrimRight(baseURL, "/"),
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
	for _, o := range opts {
		o(c)
	}
	return c
}

// SetToken sets the bearer token used on subsequent requests.
func (c *Client) SetToken(token string) {
	c.mu.Lock()
	c.token = token
	c.mu.Unlock()
}

func (c *Client) currentToken() string {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.token
}

// APIError is returned for any non-2xx response from arc-server.
type APIError struct {
	Status int
	Method string
	Path   string
	Body   string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("arc-server %s %s failed (%d): %s", e.Method, e.Path, e.Status, e.Body)
}

// ----- Auth -----

// LoginResponse is the payload of POST /v1/auth/<mount>/login.
type LoginResponse struct {
	Data struct {
		Token           string   `json:"token"`
		IdentityID      string   `json:"identityId"`
		Policies        []string `json:"policies"`
		TokenTTLSeconds int      `json:"tokenTtlSeconds"`
	} `json:"data"`
}

// LoginKubernetes authenticates with a Kubernetes ServiceAccount token via the named auth
// mount + role, caches the returned arc JWT, and returns it. `mount` defaults to
// "kubernetes" when empty.
func (c *Client) LoginKubernetes(ctx context.Context, mount, role, saToken string) (*LoginResponse, error) {
	if mount == "" {
		mount = "kubernetes"
	}
	body := map[string]string{"role": role, "jwt": saToken}
	var out LoginResponse
	if err := c.do(ctx, http.MethodPost, "auth/"+mount+"/login", body, &out, withoutAuth()); err != nil {
		return nil, err
	}
	c.SetToken(out.Data.Token)
	return &out, nil
}

// ----- KV v2 -----

// KVSecret is the response of GET /v1/<mount>/data/<path>.
type KVSecret struct {
	Data struct {
		Data     map[string]any `json:"data"`
		Metadata struct {
			Version int `json:"version"`
		} `json:"metadata"`
	} `json:"data"`
}

// KVGet reads a versioned KV v2 secret. `mount` defaults to "secret" when empty. Pass
// version <= 0 for the latest version.
func (c *Client) KVGet(ctx context.Context, mount, path string, version int) (*KVSecret, error) {
	if mount == "" {
		mount = "secret"
	}
	p := fmt.Sprintf("%s/data/%s", strings.Trim(mount, "/"), strings.TrimLeft(path, "/"))
	if version > 0 {
		p += "?version=" + fmt.Sprint(version)
	}
	var out KVSecret
	if err := c.do(ctx, http.MethodGet, p, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// KVPut writes a KV v2 secret (previous versions are retained).
func (c *Client) KVPut(ctx context.Context, mount, path string, data map[string]any) error {
	if mount == "" {
		mount = "secret"
	}
	p := fmt.Sprintf("%s/data/%s", strings.Trim(mount, "/"), strings.TrimLeft(path, "/"))
	return c.do(ctx, http.MethodPost, p, map[string]any{"data": data}, nil)
}

// ----- Dynamic credentials -----

// DynamicCredential is the response of GET /v1/<mount>/creds/<role>.
type DynamicCredential struct {
	Data          map[string]any `json:"data"`
	LeaseID       string         `json:"lease_id"`
	LeaseDuration int            `json:"lease_duration"`
	Renewable     bool           `json:"renewable"`
}

// IssueDynamic mints a short-lived dynamic credential from a plugin-backed mount (aws, gcp,
// github-app, database, …). Pass ttlSeconds <= 0 to use the role default.
func (c *Client) IssueDynamic(ctx context.Context, mount, role string, ttlSeconds int) (*DynamicCredential, error) {
	p := fmt.Sprintf("%s/creds/%s", strings.Trim(mount, "/"), strings.TrimLeft(role, "/"))
	if ttlSeconds > 0 {
		p += "?ttl=" + fmt.Sprint(ttlSeconds)
	}
	var out DynamicCredential
	if err := c.do(ctx, http.MethodGet, p, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RevokeLease revokes a lease by id (POST /v1/sys/leases/revoke).
func (c *Client) RevokeLease(ctx context.Context, leaseID string) error {
	return c.do(ctx, http.MethodPost, "sys/leases/revoke", map[string]string{"lease_id": leaseID}, nil)
}

// ----- internals -----

type reqOpts struct{ noAuth bool }

func withoutAuth() func(*reqOpts) { return func(o *reqOpts) { o.noAuth = true } }

// do issues a request to /v1/<path>, JSON-encoding body (if any) and decoding into out (if
// non-nil). On a 401 it clears the cached token and retries once.
func (c *Client) do(ctx context.Context, method, path string, body, out any, opts ...func(*reqOpts)) error {
	var o reqOpts
	for _, fn := range opts {
		fn(&o)
	}

	resp, err := c.send(ctx, method, path, body, o)
	if err != nil {
		return err
	}
	if resp.StatusCode == http.StatusUnauthorized && !o.noAuth {
		// Cached token may have been revoked server-side; drop it and retry once.
		_ = resp.Body.Close()
		c.SetToken("")
		resp, err = c.send(ctx, method, path, body, o)
		if err != nil {
			return err
		}
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return &APIError{Status: resp.StatusCode, Method: method, Path: path, Body: string(raw)}
	}
	if out != nil && len(raw) > 0 {
		if err := json.Unmarshal(raw, out); err != nil {
			return fmt.Errorf("arc-server %s %s: decode response: %w", method, path, err)
		}
	}
	return nil
}

func (c *Client) send(ctx context.Context, method, path string, body any, o reqOpts) (*http.Response, error) {
	u := c.baseURL + "/v1/" + strings.TrimLeft(path, "/")
	if _, err := url.Parse(u); err != nil {
		return nil, fmt.Errorf("invalid url %q: %w", u, err)
	}

	var rdr io.Reader
	if body != nil {
		buf, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("encode request body: %w", err)
		}
		rdr = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, method, u, rdr)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if !o.noAuth {
		if tok := c.currentToken(); tok != "" {
			req.Header.Set("Authorization", "Bearer "+tok)
		}
	}
	return c.httpClient.Do(req)
}

File: crates/arc-agent/src/arc.rs
//! Thin client to arc-server's REST API. Owns the lifecycle of the arc JWT — logs in via
//! the configured auth method on first use and refreshes ahead of expiry.
//!
//! Trust model: this client never decides authorization. arc-server's `JwtAuthGuard` +
//! `CapabilityGuard` (against `@arc/grants`) gate every call; a 403 propagates as an error
//! to the caller without retry. On 401 we drop the cached token and retry once — that
//! covers the case where the token was revoked server-side (policy detach + cache miss).

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::sync::Mutex;
use zeroize::Zeroize;

use crate::config::AuthConfig;

/// JWT-clearing wrapper so the cached token isn't left lying around in process memory longer
/// than necessary. We don't zeroize on every clone (that'd defeat the purpose); we zeroize
/// when the cache entry is replaced.
struct CachedJwt {
    token: String,
    expires_at: Instant,
}

impl Drop for CachedJwt {
    fn drop(&mut self) {
        self.token.zeroize();
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LoginResponse {
    pub data: LoginData,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LoginData {
    pub token: String,
    #[serde(default)]
    pub identity_id: String,
    #[serde(default)]
    pub policies: Vec<String>,
    pub token_ttl_seconds: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KvResponse {
    pub data: KvBody,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KvBody {
    pub data: serde_json::Value,
    #[serde(default)]
    pub metadata: KvMetadata,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct KvMetadata {
    #[serde(default)]
    pub version: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DynamicCredsResponse {
    pub data: serde_json::Value,
    pub lease_id: String,
    pub lease_duration: u64,
    #[serde(default)]
    pub renewable: bool,
}

#[derive(Debug, thiserror::Error)]
pub enum ArcError {
    #[error("arc-server HTTP error: {0}")]
    Http(#[from] reqwest::Error),
    #[error("arc-server returned {status}: {body}")]
    Server { status: u16, body: String },
    #[error("auth: failed to read token file: {0}")]
    TokenIo(#[from] std::io::Error),
    #[error("auth: not yet configured (call login() first)")]
    NotAuthenticated,
}

/// Hook used by tests to swap out the source of the SA token (or the OIDC ID token, when we
/// grow that auth type).
#[async_trait::async_trait]
pub trait TokenSource: Send + Sync {
    async fn read_token(&self) -> Result<String, ArcError>;
}

pub struct FileTokenSource {
    pub path: std::path::PathBuf,
}

#[async_trait::async_trait]
impl TokenSource for FileTokenSource {
    async fn read_token(&self) -> Result<String, ArcError> {
        let raw = tokio::fs::read_to_string(&self.path).await?;
        Ok(raw.trim().to_string())
    }
}

#[derive(Clone)]
pub struct ArcClient {
    base_url: String,
    auth: AuthConfig,
    http: reqwest::Client,
    token_source: Arc<dyn TokenSource>,
    cache: Arc<Mutex<Option<CachedJwt>>>,
    refresh_lead: Duration,
}

impl ArcClient {
    pub fn new(
        base_url: impl Into<String>,
        auth: AuthConfig,
        token_source: Arc<dyn TokenSource>,
    ) -> Self {
        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("reqwest client");
        Self {
            base_url: base_url.into(),
            auth,
            http,
            token_source,
            cache: Arc::new(Mutex::new(None)),
            refresh_lead: Duration::from_secs(60),
        }
    }

    /// Lets tests construct the client with an injectable HTTP client + clock.
    pub fn new_with_http(
        base_url: impl Into<String>,
        auth: AuthConfig,
        token_source: Arc<dyn TokenSource>,
        http: reqwest::Client,
    ) -> Self {
        Self {
            base_url: base_url.into(),
            auth,
            http,
            token_source,
            cache: Arc::new(Mutex::new(None)),
            refresh_lead: Duration::from_secs(60),
        }
    }

    /// Force a login. Returns the new arc JWT.
    pub async fn login(&self) -> Result<String, ArcError> {
        let token = match &self.auth {
            AuthConfig::Kubernetes { mount, role, .. } => {
                self.login_kubernetes(mount, role).await?
            }
        };
        Ok(token)
    }

    async fn login_kubernetes(&self, mount: &str, role: &str) -> Result<String, ArcError> {
        let sa_token = self.token_source.read_token().await?;
        let url = format!("{}/v1/auth/{}/login", self.base_url.trim_end_matches('/'), mount);
        let body = serde_json::json!({ "role": role, "jwt": sa_token });
        let res = self
            .http
            .post(&url)
            .header(reqwest::header::CONTENT_TYPE, "application/json")
            .json(&body)
            .send()
            .await?;
        let status = res.status();
        if !status.is_success() {
            let body = res.text().await.unwrap_or_default();
            return Err(ArcError::Server {
                status: status.as_u16(),
                body,
            });
        }
        let parsed: LoginResponse = res.json().await?;
        let ttl = Duration::from_secs(parsed.data.token_ttl_seconds);
        let cached = CachedJwt {
            token: parsed.data.token.clone(),
            expires_at: Instant::now() + ttl,
        };
        let mut guard = self.cache.lock().await;
        *guard = Some(cached);
        Ok(parsed.data.token)
    }

    /// Return a non-expired arc JWT, logging in if necessary.
    async fn jwt(&self) -> Result<String, ArcError> {
        let now = Instant::now();
        {
            let guard = self.cache.lock().await;
            if let Some(entry) = guard.as_ref() {
                if entry.expires_at.saturating_duration_since(now) > self.refresh_lead {
                    return Ok(entry.token.clone());
                }
            }
        }
        self.login().await
    }

    pub async fn forget_token(&self) {
        let mut guard = self.cache.lock().await;
        *guard = None;
    }

    /// `GET /v1/<mount>/data/<path>` — KV v2 read.
    pub async fn kv_get(
        &self,
        mount: &str,
        path: &str,
        version: Option<u64>,
    ) -> Result<KvResponse, ArcError> {
        let mut url = format!(
            "{}/v1/{}/data/{}",
            self.base_url.trim_end_matches('/'),
            mount.trim_matches('/'),
            path.trim_start_matches('/')
        );
        if let Some(v) = version {
            url.push_str(&format!("?version={}", v));
        }
        self.authed_json("GET", &url, None).await
    }

    /// `GET /v1/<mount>/creds/<role>` — issue dynamic credentials.
    pub async fn issue_dynamic(
        &self,
        mount: &str,
        role: &str,
        ttl_seconds: Option<u64>,
    ) -> Result<DynamicCredsResponse, ArcError> {
        let mut url = format!(
            "{}/v1/{}/creds/{}",
            self.base_url.trim_end_matches('/'),
            mount.trim_matches('/'),
            role.trim_start_matches('/')
        );
        if let Some(t) = ttl_seconds {
            url.push_str(&format!("?ttl={}", t));
        }
        self.authed_json("GET", &url, None).await
    }

    /// `POST /v1/sys/leases/revoke` — best-effort revoke.
    pub async fn revoke_lease(&self, lease_id: &str) -> Result<(), ArcError> {
        let url = format!("{}/v1/sys/leases/revoke", self.base_url.trim_end_matches('/'));
        let body = serde_json::json!({ "lease_id": lease_id });
        let _: serde_json::Value = self.authed_json("POST", &url, Some(body)).await?;
        Ok(())
    }

    async fn authed_json<T: for<'de> Deserialize<'de>>(
        &self,
        method: &str,
        url: &str,
        body: Option<serde_json::Value>,
    ) -> Result<T, ArcError> {
        let res = self.do_authed_request(method, url, body.clone()).await?;
        let status = res.status();
        // One retry on 401: cached token may have been revoked.
        if status == reqwest::StatusCode::UNAUTHORIZED {
            self.forget_token().await;
            let retry = self.do_authed_request(method, url, body).await?;
            return parse_or_err(retry).await;
        }
        parse_or_err(res).await
    }

    async fn do_authed_request(
        &self,
        method: &str,
        url: &str,
        body: Option<serde_json::Value>,
    ) -> Result<reqwest::Response, ArcError> {
        let jwt = self.jwt().await?;
        let mut req = self
            .http
            .request(reqwest::Method::from_bytes(method.as_bytes()).unwrap(), url)
            .bearer_auth(&jwt)
            .header(reqwest::header::ACCEPT, "application/json");
        if let Some(b) = body {
            req = req.json(&b);
        }
        Ok(req.send().await?)
    }
}

async fn parse_or_err<T: for<'de> Deserialize<'de>>(
    res: reqwest::Response,
) -> Result<T, ArcError> {
    let status = res.status();
    let body = res.text().await.unwrap_or_default();
    if !status.is_success() {
        return Err(ArcError::Server {
            status: status.as_u16(),
            body,
        });
    }
    if body.is_empty() {
        // Coerce empty body into T = serde_json::Value::Null for the revoke path.
        let null = serde_json::Value::Null;
        return Ok(serde_json::from_value::<T>(null).map_err(|_| ArcError::Server {
            status: 200,
            body: "empty body".into(),
        })?);
    }
    Ok(serde_json::from_str::<T>(&body).map_err(|e| ArcError::Server {
        status: status.as_u16(),
        body: format!("parse error: {} (body: {})", e, body),
    })?)
}
